lib/filterx/func-str.c

```diff
@@ -0,0 +1,459 @@
+typedef struct _FilterXExprAffix
+{
+  FilterXFunction super;
+  gboolean ignore_case;
+  FilterXExpr *haystack;
+  struct
+  {
+    FilterXExpr *expr;
+    GPtrArray *cached_strings;
+  } needle;
+  FilterXExprAffixProcessFunc process;
+} FilterXExprAffix;
+
+static GString *
+_format_str_obj(FilterXObject *obj)
+{
+  GString *str = scratch_buffers_alloc();
+
+  if(!filterx_object_repr(obj, str))
+    return NULL;
+
+  return str;
+}
+
+static gboolean
+_do_casefold(GString *str)
+{
+  if(str->len > G_MAXSSIZE)
+    return FALSE;
+
+  gchar *casefolded_str = g_utf8_casefold(str->str, (gssize) str->len);
+  g_string_assign(str, casefolded_str);
+
+  g_free(casefolded_str);
+  return TRUE;
+}
+
+static GString *
+_expr_format(FilterXExpr *expr, gboolean ignore_case)
+{
+  FilterXObject *obj = filterx_expr_eval_typed(expr);
+  if (!obj)
+    return NULL;
+
+  GString *result = _format_str_obj(obj);
+  if (!result)
+    filterx_eval_push_error("failed to extract string value, repr() failed", expr, obj);
+  filterx_object_unref(obj);
+
+  if (ignore_case && !_do_casefold(result))
+    return NULL;
+
+  return result;
+}
+
+static GString *
+_obj_format(FilterXObject *obj, gboolean ignore_case)
+{
+  GString *result = _format_str_obj(obj);
+  if (!result)
+    filterx_eval_push_error("failed to extract string value, repr() failed", NULL, obj);
+  filterx_object_unref(obj);
+
+  if (ignore_case && !_do_casefold(result))
+    return NULL;
+
+  return result;
+}
+
+static gboolean
+_string_with_cache_fill_cache(FilterXStringWithCache *self, gboolean ignore_case)
+{
+  if(!filterx_expr_is_literal(self->expr))
+    return TRUE;
+
+  return !!_expr_format(self->expr, ignore_case);
+}
+
+static FilterXStringWithCache *
+_string_with_cache_new(FilterXExpr *expr, gboolean ignore_case)
+{
+  FilterXStringWithCache *self = g_new0(FilterXStringWithCache, 1);
+  self->expr = expr;
+
+  if (!_string_with_cache_fill_cache(self, ignore_case))
+    return NULL;
+
+  return self;
+}
+
+static gboolean
+_string_with_cache_get_string_value(FilterXStringWithCache *self, gboolean ignore_case, const gchar **dst,
+                                    gsize *len)
+{
+  if(self->str_value)
+    {
+      *dst = self->str_value;
+      *len = self->str_len;
+      return TRUE;
+    }
+
+  GString *res = _expr_format(self->expr, ignore_case);
+  if (!res)
+    return FALSE;
+
+  *dst = res->str;
+  *len = res->len;
+  return TRUE;
+}
+
+static gboolean
+_cache_needle(guint64 index, FilterXExpr *value, gpointer user_data)
+{
+  gboolean *ignore_case = ((gpointer *)user_data)[0];
+  GPtrArray *cached_strings = ((gpointer *) user_data)[1];
+
+  FilterXStringWithCache *obj_with_cache =  _string_with_cache_new(value, *ignore_case);
+  if(!obj_with_cache)
+    return FALSE;
+
+  g_ptr_array_add(cached_strings, obj_with_cache);
+  return TRUE;
+}
+
+
+static gboolean
+_expr_affix_init_needle(FilterXExprAffix *self, FilterXExpr *needle)
+{
+  self->needle.expr = filterx_expr_ref(needle);
```

Comment:
You already have a reference to this expr, given by this call, no need to ref it again:
```c
FilterXExpr *haystack = filterx_function_args_get_expr(args, 0);
```